### ISSUE TITLE
VM uses more memory than allocated during OS install

### ISSUE THREAD

**Issue by aptsurfer** [cockpit-project/cockpit#15112]
Cockpit version: 215
OS: Ubuntu 20.04 LTS
Page: Virtual Machines

When creating a VM and installing the OS onto the VM, the VM uses more memory than is allocated to it.

Steps to reproduce

1. Create a virtual machine (Ubuntu 20.04 LTS, 20 GiB storage, 3 GiB RAM)
2. Monitor the memory usage as the virtual machine is being installed
3. Memory usage is highest on the "Select and install software" step.
4. Notice that more than 3 GiB RAM is in use for that VM. I observed up to 4 GiB in use.

![cockpit_extra_memory](https://user-images.githubusercontent.com/69405481/103960833-0a2c3b00-514b-11eb-9901-973607be4ffa.png)

Is this expected behavior? The recommended memory requirement for Ubuntu is 1 GiB according to [Ubuntu Server docs](https://ubuntu.com/server/docs/installation).

**Comment by KKoukiou** [comment]
This is indeed unexpected. Can you specify what is the 'Max memory' value for this VM? You can find it in the Overview tab, when clicking on the memory value link.
![Screen Shot 2021-01-11 at 08 50 38](https://user-images.githubusercontent.com/14921356/104157204-f999f000-53ea-11eb-9c22-84b987fe29cc.png)
And please verify that the values you see in the UI match also the XML value, with mapping as seen in the screenshot.

**Comment by aptsurfer** [comment]
Not sure if this is expected either, but the VM system reports less memory usage than the dashboard:

Dashboard:
![vm_highmem01](https://user-images.githubusercontent.com/69405481/104217873-d7789000-5433-11eb-9939-bb4df53fa60c.png)

VM:
![vm_highmem02](https://user-images.githubusercontent.com/69405481/104217880-db0c1700-5433-11eb-9022-c0a72a3a966f.png)

**Comment by KKoukiou** [comment]
What you see as `used` inside the VM using 'free' command and what you see in the charts outside are expected to differ which is what you notice here as well.

The VM memory usage chart shows two values.
* The used memory comes from `virsh dommemstat` command, see `rss` - Resident Set Size of the running domain's process
* And the total is the currentMemory as seen in the XML that you can configure from the UI as well.

This is (I believe) what we want to see from the host, so the actual value that that qemu process which represents the VM is using as a whole, and not what the VM reports as usage from the inside.
Naively said, the Virtualization layer that excapsulates the VM also uses memory, it's not just the guest processes (inside the VM) that use memory. However from inside the guest, there is not awareness of all the processes that is running the virtualization laayer. So that's 100% expected.

Coming back to your original question with which you reported the issue, I that's more unexpected, what you can do next time you notice it is run:
`$ virsh dommemstat your-vm-name 
actual 3145728
last_update 0
rss 64984 <------------------- and notice that the value of this (which is KB) matches the value that you see as used in the chart, of course converted to the displayed unit
``` 
If we really display the correct RSS value in the charts, I would propose that you ask in `libivirt` project why this might be happening, otherwise we will debug more here.

**Comment by KKoukiou** [comment]
@aptsurfer in any case your issue made me understand that from the UI it's not clear what the charts memory usage actual represents, and that it's not the same as guest reported usage. I will try to add documentation links to the charts labels to explain this better.

**Comment by aptsurfer** [comment]
Thanks! Documentation really helps people like me. It also makes sense that it would report the memory usage of the VM as a whole.

Here is a different example:
![vm_mem03](https://user-images.githubusercontent.com/69405481/104386471-2a823e00-552d-11eb-91ee-8ff93c1c39c3.png)

```
actual 1048576
swap_in 0
swap_out 0
major_fault 0
minor_fault 3342
unused 882776
available 1009248
usable 815680
last_update 1610492521
disk_caches 99384
hugetlb_pgalloc 0
hugetlb_pgfail 0
rss 1443476
```

Seems like the RSS value is correct.

**Comment by KKoukiou** [comment]
So the last graph can be explained by the following points:

*  The qemu process uses more memory than what it makes available to the guest. The memory qemu is using is the value 'rss' here (1.4GB) but the amount that's available to the guest is the `available` entry which is one GB here (\available is the amount of usable memory as seen by the domain (in KiB)). So seeing RSS > currentMemory can be expected, since the currentMemory is referring to max memory as seen by the guest.

* When these are forked, their memory is also copied, but with copy-on-write. RSS counts with copied memory pages twice, once for the parent and once for the child. So there can be a lot of shared memory, which is however counted twice (for a single forked process) in the RSS values that we display, making the sum or RSS memory values, bigger than the total available memory.

* ?? How RSS can be bigger than the host available memory remains still a mystery to me. If you 're interested I suggest you ask someone from libvirt project, or open a bugzilla about it. But this is outside of my understanding and the scope of this project.



To sum up, I will keep this issue open, as a reminder to add inline documentation in the charts, and maybe rethink the values that we display there, if they should be from the perpective of the host of the guest.

**Comment by aptsurfer** [comment]
To clarify (to your third bullet point), the host had more than 4 GiB of available memory in the first example. So I guess I should just be aware that qemu might take a little more memory than what I give the virtual machine. Thanks for helping.

I think having the host perspective is still of value, though it might be nice to see if the VM is maxing out its internal memory usage too.

**Comment by KKoukiou** [comment]
> To clarify (to your third bullet point), the host had more than 4 GiB of available memory in the first example. So I guess I should just be aware that qemu might take a little more memory than what I give the virtual machine. Thanks for helping.
> 
> I think having the host perspective is still of value, though it might be nice to see if the VM is maxing out its internal memory usage too.

Now it makes sense! super, I will rethink how to display these values! Thanks for the feedback :)
